test-sbat.c

```diff
@@ -0,0 +1,876 @@
+	struct sbat_var *test_entry;
+	for (i = 0; i < num_elem; i++) {
+		test_entry = create_mock_sbat_var_entry(var_array[i].component_name,
+							var_array[i].component_generation);
+		if (!test_entry)
+			return NULL;
+		list_add(&test_entry->list, test_sbat_entries);
+	}
+	return test_sbat_entries;
+}
+
+void
+free_mock_sbat_entries(list_t *entries)
+{
+	list_t *pos = NULL;
+	list_t *n = NULL;
+	struct sbat_var *entry;
+
+	if (entries)
+	{
+		list_for_each_safe(pos, n, entries)
+		{
+			entry = list_entry(pos, struct sbat_var, list);
+			list_del(&entry->list);
+			if (entry->component_name)
+				FreePool((CHAR8 *)entry->component_name);
+			if (entry->component_generation)
+				FreePool((CHAR8 *)entry->component_generation);
+			FreePool(entry);
+		}
+		FreePool(entries);
+	}
+}
+
+/*
+ * parse_sbat() tests
+ */
+void
+test_parse_sbat_null_sbat_base(void)
+{
+	size_t sbat_size = 20;
+	struct sbat sbat = { 0, NULL };
+	EFI_STATUS status;
+
+	status = parse_sbat(NULL, sbat_size, &sbat);
+
+	assert(status == EFI_INVALID_PARAMETER);
+}
+
+void
+test_parse_sbat_zero_sbat_size(void)
+{
+	char sbat_base[] = "test1,1,SBAT test1,acme,1,testURL\n";
+	size_t sbat_size = 0;
+	struct sbat sbat = { 0, NULL };
+	EFI_STATUS status;
+
+	status = parse_sbat(sbat_base, sbat_size, &sbat);
+
+	assert(status == EFI_INVALID_PARAMETER);
+}
+
+void
+test_parse_sbat_null_sbat(void)
+{
+	char sbat_base[] = "test1,1,SBAT test1,acme,1,testURL\n";
+	size_t sbat_size = sizeof(sbat_base);
+	EFI_STATUS status;
+
+	status = parse_sbat(sbat_base, sbat_size, NULL);
+
+	assert(status == EFI_INVALID_PARAMETER);
+}
+
+void
+test_parse_sbat_no_newline(void)
+{
+	char sbat_base[] = "test1,1,SBAT test1,acme,1,testURL";
+	size_t sbat_size = sizeof(sbat_base);
+	struct sbat sbat = { 0 };
+	EFI_STATUS status;
+
+	status = parse_sbat(sbat_base, sbat_size, &sbat);
+
+	assert(status == EFI_INVALID_PARAMETER);
+}
+
+void
+test_parse_sbat_too_few_elem(void)
+{
+	char sbat_base[] = "test1,1,SBAT test1,acme,1\n";
+	size_t sbat_size = sizeof(sbat_base);
+	struct sbat sbat = { 0 };
+	EFI_STATUS status;
+
+	status = parse_sbat(sbat_base, sbat_size, &sbat);
+
+	assert(status == EFI_INVALID_PARAMETER);
+}
+
+void
+test_parse_sbat_too_many_elem(void)
+{
+	char sbat_base[] = "test1,1,SBAT test1,acme,1,testURL,testURL2\n";
+	size_t sbat_size = sizeof(sbat_base);
+	struct sbat sbat = { 0 };
+	EFI_STATUS status;
+
+	status = parse_sbat(sbat_base, sbat_size, &sbat);
+
+	assert(status == EFI_SUCCESS);
+	assert(sbat.size == 1);
+	assert(strncmp(sbat.entries[0]->component_name, "test1", MAX_SIZE) == 0);
+	assert(strncmp(sbat.entries[0]->component_generation, "1", MAX_SIZE) == 0);
+	assert(strncmp(sbat.entries[0]->vendor_name, "SBAT test1", MAX_SIZE) == 0);
+	assert(strncmp(sbat.entries[0]->vendor_package_name, "acme", MAX_SIZE) == 0);
+	assert(strncmp(sbat.entries[0]->vendor_version, "1", MAX_SIZE) == 0);
+	assert(strncmp(sbat.entries[0]->vendor_url, "testURL,testURL2", MAX_SIZE) == 0);
```

Comment:
When `.sbat` is malformed with too many elements, should we expect `EFI_SUCCESS` and the `vendor_url` to contain everything between 5th `,` and `\n`?